Editor/assets/shaders/phong.glsl

```diff
@@ -50,21 +68,87 @@ vec4 pointLightCalculate(PointLight light, vec3 norm)
 	float diff = max(dot(norm, lightDir), 0.0);
 	vec4 diffuse = diff * light.color;
 
-	return diffuse * attentuation;
+	//specular
+	vec3 reflectDir = normalize(reflect(-lightDir,norm));
+	float spec = pow(max(dot(reflectDir,viewDir),0.0),32);
+	vec4 specular = spec * light.color;
+	diffuse = diffuse * attentuation;
+	specular = specular*attentuation;
+	vec4 result = diffuse + specular;
+	return result;
 }
 
-void main()
+vec4 directionalLightCalculate(DirectionalLight light,vec3 norm,vec3 viewDir)
+{	
+	vec4 color =  vec4(light.color);
+	vec3 direction = normalize(vec3(light.direction));
+	float diff = max(dot(norm, direction), 0.0);
+	vec4 diffuse = diff * color;
+
+
+	//specular
+	vec3 reflectDir = normalize(reflect(-direction,norm));
+	float spec = pow(max(dot(reflectDir,viewDir),0.0),32);
+	vec4 specular = spec * color;
+
+	vec4 result = diffuse +specular;
+	return result;
+}
+
+vec4 spotLightCalculate(SpotLight light, vec3 norm,vec3 viewDir)
 {
+
+	vec3 lightDir = normalize(light.position.xyz - v_worldPos);
+	vec4 result;
+
+	
+
+	float theta = dot(lightDir,light.direction.xyz);
+	if(theta > light.outerCutoff)
+	{	
+		float intensity = clamp((theta - light.outerCutoff) / (light.innerCutoff - light.outerCutoff), 0.0, 1.0);
+		float diff = max(dot(norm, lightDir), 0.0);
+		vec4 diffuse = diff * light.color;
+		vec3 reflectDir = normalize(reflect(-lightDir,norm));
+		float spec = pow(max(dot(reflectDir,viewDir),0.0),64);
+		vec4 specular = spec * light.color;
+		diffuse *=   intensity;
+		specular *= intensity;
+		result = diffuse + specular;
+	}
+	else
+	{
+		result = vec4(0,0,0,0);
+	}
+	
+	
+	
+	return result;
+}
+
+
+void main()
+{	
+	
 	vec3 norm = normalize(v_normal);
+	vec3 viewDir = normalize(cameraPosition - v_worldPos);
 	color = vec4(0.3, 0.3, 0.3, 1.0);
 	// for (int i = 0; i < u_numPointLights; i++)
 	// {
 	// 	color += vec4(pointLightCalculate(u_pointLights[i]), 0.0);
 	// }
-	color += pointLightCalculate(u_pointLights[0], norm);
-	color += pointLightCalculate(u_pointLights[1], norm);
-	color += pointLightCalculate(u_pointLights[2], norm);
-	color += pointLightCalculate(u_pointLights[3], norm);
+	color += pointLightCalculate(u_pointLights[0], norm,viewDir);
+	color += pointLightCalculate(u_pointLights[1], norm,viewDir);
+	color += pointLightCalculate(u_pointLights[2], norm,viewDir);
+	color += pointLightCalculate(u_pointLights[3], norm,viewDir);
+	color += spotLightCalculate(u_spotLights[0], norm,viewDir);
+	color +=  spotLightCalculate(u_spotLights[1], norm,viewDir);
```

Comment:
Try to put spaces after commas, and keep an eye out for double spaces like the ones here.